federationsender/queue/destinationqueue.go

```diff
@@ -95,11 +94,13 @@ func (oq *destinationQueue) sendEvent(nid int64) {
 	// We've successfully added a PDU to the transaction so increase
 	// the counter.
 	oq.transactionCount.Add(1)
-	// Signal that we've sent a new PDU. This will cause the queue to
-	// wake up if it's asleep. The return to the Add function will only
-	// be 1 if the previous value was 0, e.g. nothing was waiting before.
-	if oq.pendingPDUs.Add(1) == 1 {
-		oq.notifyPDUs <- true
+	// Wake up the queue if it's asleep.
+	oq.wakeQueueIfNeeded()
+	// If we're blocking on waiting PDUs then tell the queue that we
+	// have work to do.
+	select {
+	case oq.notifyPDUs <- true:
```

Comment:
We always want to do `wakeQueueIfNeeded` and doing a soft-push onto the notify channel so I would fold this into a single function (and have a function args for whether this is a PDU/EDU/invite)

Reply:
When EDUs and invites are also following the same mechanism then that would probably make sense. For now it doesn't as `incomingEDUs`/`incomingInvites` are used instead.